src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/WebSocketDeflater.cs

```diff
@@ -0,0 +1,229 @@
+            const int MaxInitialBufferLength = 1024 * 1024;
+
+            // For small payloads there might actually be overhead in the compression and the resulting
+            // output might be larger than the payload. This is why we rent at least 4KB initially.
+            const int MinInitialBufferLength = 4 * 1024;
+
+            _buffer = ArrayPool<byte>.Shared.Rent(Math.Min(Math.Max(payload.Length, MinInitialBufferLength), MaxInitialBufferLength));
+            int position = 0;
+
+            while (true)
+            {
+                DeflatePrivate(payload, _buffer.AsSpan(position), continuation, endOfMessage,
+                    out int consumed, out int written, out bool needsMoreOutput);
+                position += written;
+
+                if (!needsMoreOutput)
+                {
+                    break;
+                }
+
+                payload = payload.Slice(consumed);
+
+                // Rent a 30% bigger buffer
+                byte[] newBuffer = ArrayPool<byte>.Shared.Rent((int)(_buffer.Length * 1.3));
+                _buffer.AsSpan(0, position).CopyTo(newBuffer);
+                ArrayPool<byte>.Shared.Return(_buffer);
+                _buffer = newBuffer;
+            }
+
+            return new ReadOnlySpan<byte>(_buffer, 0, position);
+        }
+
+        private void DeflatePrivate(ReadOnlySpan<byte> payload, Span<byte> output, bool continuation, bool endOfMessage,
+            out int consumed, out int written, out bool needsMoreOutput)
+        {
+            Debug.Assert(!continuation || _stream is not null, "Invalid state. The stream should not be null in continuations.");
+
+            if (_stream is null)
+            {
+                Initialize();
+            }
+
+            UnsafeDeflate(payload, output, out consumed, out written, out needsMoreOutput);
+            if (needsMoreOutput)
+            {
+                return;
+            }
+
+            // See comment by Mark Adler https://github.com/madler/zlib/issues/149#issuecomment-225237457
+            // At that point there will be at most a few bits left to write.
+            // Then call deflate() with Z_FULL_FLUSH and no more input and at least six bytes of available output.
+            if (output.Length - written < 6)
+            {
+                needsMoreOutput = true;
+                return;
+            }
+            written += UnsafeFlush(output.Slice(written));
+            Debug.Assert(output.Slice(written - WebSocketInflater.FlushMarkerLength, WebSocketInflater.FlushMarkerLength)
+                               .EndsWith(WebSocketInflater.FlushMarker), "The deflated block must always end with a flush marker.");
+
+            if (endOfMessage)
+            {
+                // As per RFC we need to remove the flush markers
+                written -= WebSocketInflater.FlushMarkerLength;
+            }
+
+            if (endOfMessage && !_persisted)
+            {
+                _stream.Dispose();
```

Comment:
As far as I know, this isn't possible. This `_stream` object is a `SafeHandle` which when disposed releases the native resources (read zlib). The only way to reinitialize it, is to create it again.

Reply:
In zlib itself I actually see the methods which seem to do what is needed

`ZEXTERN int ZEXPORT deflateReset OF((z_streamp strm));`
> This function is equivalent to deflateEnd followed by deflateInit, but does not free and reallocate the internal compression state. The stream will leave the compression level and any other attributes that may have been set unchanged.

`ZEXTERN int ZEXPORT inflateReset OF((z_streamp strm));`
> This function is equivalent to inflateEnd followed by inflateInit, but does not free and reallocate the internal decompression state. The The stream will keep attributes that may have been set by inflateInit2.

Should we maybe add them to Interop.zlib.cs, as we don't have them right now? @carlossanlop

Reply:
This seems possible, however I have one concert about it. Usually when context takeover is disallowed, it's used to keep the memory footprint per socket to minimal and only compress certain messages which the user knows it's worth it. If we do this, then the memory usage will be the same with or without context takeover, even if you only ever send just one compressed message. 

See https://github.com/dotnet/runtime/issues/31088#issuecomment-539603012 for rough numbers.

Reply:
>I've looked at what chromium does, and indeed it uses what you suggest. See https://github.com/chromium/chromium/blob/99314be8152e688bafbbf9a615536bdbb289ea87/net/websockets/websocket_deflater.cc#L132-L136.
>
>I guess in the case when the websocket is not a server, we can opt into that and in the case of a server websocket, keep the current implementation.

I guess this comment was meant to be here 😄 
What would be the difference between client and server in this matter, why you think it will be ok for client, but not for server? I would expect both of them having the same behavior 🤔

Reply:
My reasoning is that when we're the client side of the websocket, we usually have one or just a few instances. 

When we're the server side, I expect that memory usage would be a concern since we can expect to have many websocket instances. This is the only reason I would ever use no context takeover (or having websockets being one sided - receiving or sending is the prevailing part).

Reply:
> Should we maybe add them to Interop.zlib.cs, as we don't have them right now? @carlossanlop

@CarnaViire I don't have a problem with having those two new methods added to Interop.zlib.cs if you need them. Let me know if you do, and I'll be happy to help review.

Reply:
It makes sense to me. @scalablecory do you have any opinion on whether we should have a different behavior for client and server here, given the arguments above?

Reply:
In the case that "takeover" is disabled, I agree that the point of that is largely to reduce idle memory usage, so we shouldn't be hanging on to a stream.

I don't think we need any new public API, but I do think it'd be very beneficial to have a pool of these objects, and to have the pool call `deflateReset` or `inflateReset` when one is returned. The pool can keep track of the last time each object was used, and have a timer that flushes objects from the pool when they haven't been used in a while.

This will allow us to reduce memory usage but still avoid tons of per-message overhead.

Reply:
When implementing this timer, be sure to avoid capturing execution context. See example here:

https://github.com/dotnet/runtime/blob/553338882229caaa892227e251a0c15fbb6b13dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L95-L142